src/lib/stream/shake_cipher/shake_cipher.h

```diff
@@ -1,77 +1,128 @@
  * (C) 2016 Jack Lloyd
- *     2022 René Meusel, Michael Boric - Rohde & Schwarz Cybersecurity
+ *     2022, 2023 René Meusel, Michael Boric - Rohde & Schwarz Cybersecurity
  *
  * Botan is released under the Simplified BSD License (see license.txt)
  */
 
 #ifndef BOTAN_SHAKE_CIPHER_H_
 #define BOTAN_SHAKE_CIPHER_H_
 
-#include <botan/secmem.h>
+#include <botan/exceptn.h>
 #include <botan/stream_cipher.h>
+#include <botan/internal/shake_xof.h>
 
 namespace Botan {
 
+namespace detail {
+
 /**
 * Base class for SHAKE-based XOFs presented as a stream cipher
+*
+* TODO: This class could potentially act as a generic adapter to
+*       use arbitrary XOFs as a basis for stream ciphers.
 */
-class SHAKE_Cipher : public StreamCipher {
-   protected:
-      explicit SHAKE_Cipher(size_t shake_rate);
 
+template <typename XofT, typename DerivedT>
+class XOF_Stream_Cipher_Adapter : public StreamCipher {
    public:
-      /**
-      * Seeking is not supported, this function will throw
-      */
-      void seek(uint64_t offset) final;
+      XOF_Stream_Cipher_Adapter() : m_keystream_buffer(buffer_size()), m_bytes_generated(0) {}
+
+      void seek(uint64_t) final { throw Not_Implemented("SHAKE_Cipher::seek"); }
+
+      void clear() final {
+         m_xof.clear();
+         zeroise(m_keystream_buffer);
+         m_bytes_generated = 0;
+      }
 
-      void clear() final;
+      Key_Length_Specification key_spec() const final { return Key_Length_Specification(1, 160); }
 
-      Key_Length_Specification key_spec() const final;
+      bool has_keying_material() const final { return !m_xof.accepts_input(); }
 
-      bool has_keying_material() const final;
+      size_t buffer_size() const final { return m_xof.block_size(); }
 
-      size_t buffer_size() const final;
+      std::string name() const override { return m_xof.name(); }
+
+      std::unique_ptr<StreamCipher> new_object() const override { return std::make_unique<DerivedT>(); }
 
    private:
-      void key_schedule(const uint8_t key[], size_t key_len) final;
-      /**
-      * Produce more XOF output
-      */
-      void cipher_bytes(const uint8_t in[], uint8_t out[], size_t length) final;
-      void generate_keystream(uint8_t out[], size_t length) override;
+      void key_schedule(const uint8_t key[], size_t key_len) final {
+         m_xof.clear();
+         m_xof.update({key, key_len});
+         m_xof.output({});  // empty output request to transition the XOF to output state
+      }
 
       /**
-      * IV not supported, this function will throw unless iv_len == 0
+      * Produce more XOF output
       */
-      void set_iv_bytes(const uint8_t iv[], size_t iv_len) final;
-
-   protected:
-      size_t m_shake_rate;
+      void cipher_bytes(const uint8_t in[], uint8_t out[], size_t length) final {
+         assert_key_material_set();
+
+         const auto block_size = m_keystream_buffer.size();
+
+         auto cipher_some = [&](size_t bytes) {
```

Comment:
I don't see why this logic has to be in a header

Reply:
Yeah, that was still sketchy. Addressed in #3675.

Now that `Keccak_Permutation::squeeze()` can handle multiple calls to pull the sponge output in arbitrarily-sized chunks, I decided not the base `SHAKE_Cipher` on the `SHAKE_XOF` but directly on `Keccak_Permutation`. There's no difference in level of abstraction anyway.